rest2html: treat referenced wrapped images in base document as inlined

jdknight opened the pull request:
docutils will only add newlines around images it believes are inlined. For images held in references, it checks the parent of the reference if its a `TextElement` to consider it inlined. Since a document is not a `TextElement` type, it will wrap an image with newlines. For GitHub output, this is not desired and will result in the extra whitespace being rendered with a reference's decorative line. To avoid this, always strip any appended suffixes for images in this scenario.

----

There are a various GitHub projects which reveal the issue. For example, [Sphinx's `README.rst`](https://github.com/sphinx-doc/sphinx/blob/master/README.rst) shows this issue:

<kbd>
<img src="https://github.com/user-attachments/assets/36248550-e0da-4d44-97eb-3961b8b1d488">
</kbd>

With the changes made in this merge request, the following shows a rendering of HTML before and after the change:

<kbd>
<img src="https://github.com/user-attachments/assets/23db7a8f-24cd-4071-8a25-5ce832d21296">
</kbd>

jdknight committed:
rest2html: treat referenced wrapped images in base document as inlined

docutils will only add newlines around images it believes are inlined.
For images held in references, it checks the parent of the reference if
its a `TextElement` to consider it inlined. Since a document is not a
`TextElement` type, it will wrap an image with newlines. For GitHub
output, this is not desired and will result in the extra whitespace
being rendered with a reference's decorative line. To avoid this,
always strip any appended suffixes for images in this scenario.

Signed-off-by: James Knight <git@jdknight.me>

jdknight commented:
!unstale

jdknight commented:
!unstale

jdknight commented:
!unstale

zkoppert commented:
Hey @jdknight — thanks for sticking with this PR, and sorry it took so long to get eyes on it.

I rebased your branch onto master to get CI running (it was 46 commits behind), and all checks pass across Ruby 3.2/3.3/3.4. The fix itself looks good — we reviewed it thoroughly and the logic is sound.

One thing we'd like to see before merging: a **regression test**. The existing test suite uses Nokogiri with `noblanks`, which normalizes whitespace-only text nodes — so the current tests wouldn't catch it if this behavior regressed. A string-level assertion (not DOM-equality) that verifies there's no trailing whitespace between the `<img>` and closing `</a>` for a top-level referenced image would do the trick.

Something like an RST fixture with a `.. image::` directive using `:target:`, and an assertion on the raw HTML output before Nokogiri parsing. Happy to help brainstorm the approach if that would be useful.

jdknight commented:
@zkoppert, there appears to be a couple of locations in `assert_html_equal`/`normalize_html` that would need to be adjusted to allow `assert_html_equal` to support strict whitespace checks, let alone finding a graceful way to opt-in to strict checks through the generic markup test discovery logic.

Maybe for this case a simple test can be appended. For example:

```
  def test_blanks_rst_inlined_reference
    expected = '<img src="https://example.com/img.svg"></a>'

    actual = GitHub::Markup.render_s(GitHub::Markups::MARKUP_RST, <<~RST
    .. image:: https://example.com/img.svg
        :target: https://example.com/
    RST
    )

    assert_equal expected, actual.strip.lines.last
  end
```

And if more tests are needed later that require strict whitespace checks, this can be looked at then.

jdknight committed:
test: validate inlined rst references do not add whitespaces

Verifying that inlined references generated by reStructuredText/docutil
do no generate whitespaces between the img tag and closing a tag.

Signed-off-by: James Knight <git@jdknight.me>

jdknight commented:
!unstale
